datafusion/core/src/physical_optimizer/enforcement.rs

```diff
@@ -0,0 +1,1739 @@
+        ) {
+            let new_join_on = if !new_positions.is_empty() {
+                new_join_conditions(&left_keys, &right_keys)
+            } else {
+                on.clone()
+            };
+            let new_options = if !new_positions.is_empty() {
+                let mut new_sort_options = vec![];
+                for idx in 0..sort_options.len() {
+                    new_sort_options.push(sort_options[new_positions[idx]])
+                }
+                new_sort_options
+            } else {
+                sort_options.clone()
+            };
+
+            let new_left = adjust_input_keys_down_recursively(left.clone(), left_keys)?;
+            let new_right =
+                adjust_input_keys_down_recursively(right.clone(), right_keys)?;
+
+            Ok(Arc::new(SortMergeJoinExec::try_new(
+                new_left,
+                new_right,
+                new_join_on,
+                *join_type,
+                new_options,
+                *null_equals_null,
+            )?))
+        } else {
+            let new_left = adjust_input_keys_down_recursively(
+                left.clone(),
+                join_key_pairs.left_keys,
+            )?;
+            let new_right = adjust_input_keys_down_recursively(
+                right.clone(),
+                join_key_pairs.right_keys,
+            )?;
+            Ok(Arc::new(SortMergeJoinExec::try_new(
+                new_left,
+                new_right,
+                on.clone(),
+                *join_type,
+                sort_options.clone(),
+                *null_equals_null,
+            )?))
+        }
+    } else if let Some(AggregateExec {
+        mode,
+        group_by,
+        aggr_expr,
+        input,
+        input_schema,
+        ..
+    }) = plan_any.downcast_ref::<AggregateExec>()
+    {
+        if parent_required.is_empty() {
+            plan.map_children(|plan| adjust_input_keys_down_recursively(plan, vec![]))
+        } else {
+            match mode {
+                AggregateMode::Final => plan.map_children(|plan| {
+                    adjust_input_keys_down_recursively(plan, vec![])
+                }),
+                AggregateMode::FinalPartitioned | AggregateMode::Partial => {
+                    let out_put_columns = group_by
+                        .expr()
+                        .iter()
+                        .enumerate()
+                        .map(|(index, (_col, name))| Column::new(name, index))
+                        .collect::<Vec<_>>();
+
+                    let out_put_exprs = out_put_columns
+                        .iter()
+                        .map(|c| Arc::new(c.clone()) as Arc<dyn PhysicalExpr>)
+                        .collect::<Vec<_>>();
+
+                    // Check whether the requirements can be satisfied by the Aggregation
+                    if parent_required.len() != out_put_exprs.len()
+                        || expr_list_eq_strict_order(&out_put_exprs, &parent_required)
```

Comment:
Here, we don't consider the optimization of join group with sharing the same hash table. Therefore, when the group satisfy the join key, we don't push down the required ordering.